src/protocols/der/base.c

```diff
@@ -0,0 +1,518 @@
+ *   Lesser General Public License for more details.
+ *
+ *   You should have received a copy of the GNU Lesser General Public
+ *   License along with this library; if not, write to the Free Software
+ *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/**
+ * $Id$
+ *
+ * @file protocols/der/decode.c
+ * @brief Functions to decode DER encoded data.
+ *
+ * @author Ethan Thompson (ethan.thompson@inkbridge.io)
+ *
+ * @copyright 2024 Inkbridge Networks SAS.
+ */
+
+#include "include/build.h"
+#include "der.h"
+#include "lib/util/table.h"
+#include "lib/util/types.h"
+#include <freeradius-devel/util/proto.h>
+#include <freeradius-devel/util/dict.h>
+#include <freeradius-devel/util/net.h>
+#include <stdbool.h>
+
+static uint32_t instance_count = 0;
+
+fr_dict_t const *dict_der;
+
+extern fr_dict_autoload_t libfreeradius_der_dict[];
+fr_dict_autoload_t	  libfreeradius_der_dict[] = { { .out = &dict_der, .proto = "der" }, { NULL } };
+
+extern fr_dict_attr_autoload_t libfreeradius_der_dict_attr[];
+fr_dict_attr_autoload_t	       libfreeradius_der_dict_attr[] = {
+	       { NULL }
+};
+
+fr_der_tag_constructed_t tag_labels[] = {
+	[FR_DER_TAG_BOOLEAN]	      = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_INTEGER]	      = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_BITSTRING]	      = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_OCTETSTRING]      = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_NULL]	      = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_OID]	      = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_ENUMERATED]	      = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_UTF8_STRING]      = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_SEQUENCE]	      = FR_DER_TAG_CONSTRUCTED,
+	[FR_DER_TAG_SET]	      = FR_DER_TAG_CONSTRUCTED,
+	[FR_DER_TAG_PRINTABLE_STRING] = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_T61_STRING]	      = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_IA5_STRING]	      = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_UTC_TIME]	      = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_GENERALIZED_TIME] = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_VISIBLE_STRING]   = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_GENERAL_STRING]   = FR_DER_TAG_PRIMATIVE,
+	[FR_DER_TAG_UNIVERSAL_STRING] = FR_DER_TAG_PRIMATIVE,
+};
+
+int fr_der_global_init(void)
+{
+	if (instance_count > 0) {
+		instance_count++;
+		return 0;
+	}
+
+	instance_count++;
+
+	if (fr_dict_autoload(libfreeradius_der_dict) < 0) {
+	fail:
+		instance_count--;
+		return -1;
+	}
+
+	if (fr_dict_attr_autoload(libfreeradius_der_dict_attr) < 0) {
+		fr_dict_autofree(libfreeradius_der_dict);
+		goto fail;
+	}
+
+	return 0;
+}
+
+void fr_der_global_free(void)
+{
+	// fr_assert(instance_count > 0);
+
+	if (--instance_count != 0) return;
+
+	fr_dict_autofree(libfreeradius_der_dict);
+}
+
+static int dict_flag_tagnum(fr_dict_attr_t **da_p, char const *value, UNUSED fr_dict_flag_parser_rule_t const *rules)
+{
+	fr_der_attr_flags_t *flags = fr_dict_attr_ext(*da_p, FR_DICT_ATTR_EXT_PROTOCOL_SPECIFIC);
+
+	flags->tagnum = (uint8_t)atoi(value);
```

Comment:
do we want sanity checks here for overflow, or for bad formats?

Admins make mistakes. An error of "you screwed up" is usually better than silently doing the wrong thing.